toolchain/semantics/semantics_ir_formatter.cpp

```diff
@@ -0,0 +1,654 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/semantics/semantics_ir_formatter.h"
+
+#include "llvm/ADT/StringExtras.h"
+#include "llvm/ADT/StringSet.h"
+#include "llvm/Support/SaveAndRestore.h"
+#include "toolchain/lexer/tokenized_buffer.h"
+#include "toolchain/parser/parse_tree.h"
+
+namespace Carbon {
+
+namespace {
+// Assigns names to nodes, blocks, and scopes in the Semantics IR.
```

Comment:
Adding a suggested edit adding TODOs here as a way for me to suggest things to consider for future work in a way that won't get lost in the sands of pull-request history, but definitely aren't interesting in this PR.

Feel free to reformat, move around, etc if you see a better place / way of documenting.

```suggestion
// Assigns names to nodes, blocks, and scopes in the Semantics IR.
//
// TODOs / future work ideas:
// - Add a documentation file for the textual format and link to the
//   naming section here.
// - Consider harmonizing `FooLiteral` vs. `foo_value` node names vs.
//   IR names.
//   - Also consider representing these as just `value` or `literal` 
//     in the IR and using the type to distinguish.
// - Add block names based on the control flow construct names (`for`,
//   `if`, `then`, `else`, ...). Either base this on the semantics or
//   just on the keywords used to build that control flow -- little
//   or no harm to getting it wrong. Tools will also want to be able
//   to query the construct that resulted in control flow.
```